src/progress/nested/subgraph.rs

```diff
@@ -491,6 +506,103 @@ impl<TOuter: Timestamp, TInner: Timestamp> Operate<TOuter> for Subgraph<TOuter,
     }
 }
 
+impl<TOuter: Timestamp, TInner: Timestamp> SubgraphBuilder<TOuter, TInner> {
+    /// Allocates a new input to the subgraph and returns the assigned index.
+    pub fn new_input(&mut self, shared_counts: Rc<RefCell<ChangeBatch<Product<TOuter, TInner>>>>) -> usize {
+        self.inputs += 1;
+        self.input_messages.push(shared_counts);
+        self.children[0].add_output();
+        self.inputs - 1
+    }
+
+    /// Allocates a new output from the subgraph and returns the assigned index.
+    pub fn new_output(&mut self) -> usize {
+        self.outputs += 1;
+        self.output_capabilities.push(MutableAntichain::new());
+        self.children[0].add_input();
+        self.outputs - 1
+    }
+
+    /// Introduces a dependence from the source to the target.
+    ///
+    /// This method does not effect data movement, but rather reveals to the progress tracking infrastructure
+    /// that messages produced by `source` should be expected to be consumed at `target`.
+    pub fn connect(&mut self, source: Source, target: Target) {
+        self.edge_stash.push((source, target));
+    }
+
+    /// Creates a new Subgraph from a channel allocator and "descriptive" indices.
+    pub fn new_from(index: usize, mut path: Vec<usize>) -> SubgraphBuilder<TOuter, TInner> {
+        path.push(index);
+
+        let children = vec![PerOperatorState::empty(path.clone())];
+
+        SubgraphBuilder {
+            name:                   "Subgraph".into(),
+            path:                   path,
+            index:                  index,
+
+            inputs:                 Default::default(),
+            outputs:                Default::default(),
+
+            children:               children,
+            child_count:            1,
+            edge_stash: vec![],
+
+            input_messages:         Default::default(),
+            output_capabilities:    Default::default(),
+        }
+    }
+
+    /// Allocates a new child identifier, for later use.
+    pub fn allocate_child_id(&mut self) -> usize {
+        self.child_count += 1;
+        self.child_count - 1
+    }
+
+    /// Adds a new child to the subgraph.
+    pub fn add_child(&mut self, child: Box<Operate<Product<TOuter, TInner>>>, index: usize, identifier: usize) {
+        self.children.push(PerOperatorState::new(child, index, self.path.clone(), identifier))
+    }
+
+    /// Now that initialization is complete, actually build a subgraph.
+    pub fn build<A: Allocate>(mut self, allocator: &mut A) -> Subgraph<TOuter, TInner> {
```

Comment:
One very appealing consequence of this signature is that all of the channels in a subgraph could be allocated at this point, in principle. That is, one can whip up a subgraph builder that describes the structure of the computation underneath it, but is agnostic to the actual numbers of workers / their current location / etc. Only when this method is called do we construct the actual connections even further down.

Maybe a different way to state: you could imagine somewhere a trait 

```rust
pub trait OperateBuilder<T: Timestamp> {
    fn build<A: Allocate>(mut self, allocator: &mut A) -> Box<Operate<T>>;
}
```

and one could in principle get all of a subgraph ready to go with operator builders in place, but with zero hooks into a running timely system (so "just data" at the moment), and then instantiate everything only once asked.

We did a similar thing in Naiad, but I think there was more of a start-up sequence there. I'll try to page back in why it was so important.